flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -540,15 +540,12 @@ private boolean performCheckpoint(CheckpointMetaData checkpointMetaData) throws
 
 		synchronized (lock) {
 			if (isRunning) {
+				checkpointState(checkpointMetaData);
 
-				// Since both state checkpointing and downstream barrier emission occurs in this
-				// lock scope, they are an atomic operation regardless of the order in which they occur.
-				// Given this, we immediately emit the checkpoint barriers, so the downstream operators
-				// can start their checkpoint work as soon as possible
+				// broadcast barriers after snapshot operators' states.
 				operatorChain.broadcastCheckpointBarrier(
-						checkpointMetaData.getCheckpointId(), checkpointMetaData.getTimestamp());
-
-				checkpointState(checkpointMetaData);
+						checkpointMetaData.getCheckpointId(), checkpointMetaData.getTimestamp()
+				);
```

Comment:
Disabling chaining does not necessarily mean that the data is sent over the network then. Flink will still try to schedule the tasks to the same machine.
Doing a benchmark is a good idea. I think we've also done it in the past and we concluded that the reentrant lock is too expensive if I remember correctly.